### PR TITLE
tests: Infrastructure for unit tests

### PR DESCRIPTION
**Pull request by MerryMage**
Extracted the test infrastructure commit from #1554.

Does not execute tests on the buildbots yet. Should this be done?

### CODE REVIEW

**Comment by Subv** [comment]
>Does not execute tests on the buildbots yet. Should this be done?
IMO yes

File: appveyor.yml [hunk 1]
```diff
@@ -22,6 +22,10 @@ before_build:
   - cmake -G "Visual Studio 14 2015 Win64" -DCITRA_USE_BUNDLED_QT=1 -DCITRA_USE_BUNDLED_SDL2=1 ..
   - cd ..
 
+build:
+  project: build/citra.sln
+  parallel: true
+
```
Comment (JayFoxRox) [inline appveyor.yml]: How is this related to testing?

Comment (MerryMage) [inline appveyor.yml]: Appveyor fails to properly autodetect the correct `.sln` file as the catch submodule itself has a few `.sln` files, hence it has to be explicitly specified.

**Comment by JayFoxRox** [comment]
LGTM so far BUT I'm afraid nobody will use it if there is not a single test defined.
So I'd prefer if there was an issue or even a PR (based on this one) which already implements some tests which could serve as example. I might even start adding unit tests myself then.

**Comment by MerryMage** [comment]
> a PR (based on this one) which already implements some tests which could serve as example

#1554 has a number of tests (for example [this one](https://github.com/MerryMage/citra/blob/jit2/src/tests/core/arm/jit_x64/fuzz_arm_branch.cpp)), although those aren't exactly your typical unit tests. I would suggest reading the [Catch tutorial](https://github.com/philsquared/Catch/blob/master/docs/tutorial.md) before use. I would also love to write some audio tests, I've actually been itching to.

A suggsted folder structure for the `src/tests/` directory is to reflect the project structure (for example, a test for `core/arm/jit_x64/instructions/branch.cpp` might appear in `src/tests/core/arm/jit_x64/fuzz_arm_branch.cpp`).

**Comment by Subv** [comment]
Could you rebase this so we can merge it once and for all? thanks!

**Comment by MerryMage** [comment]
rebased

**Comment by JayFoxRox** [comment]
Again please?

**Comment by MerryMage** [comment]
@JayFoxRox rebased.

**Comment by MerryMage** [comment]
Tests are now run on CI.

File: CMakeLists.txt [hunk 2]
```diff
@@ -255,6 +255,8 @@ endif()
 
 add_subdirectory(externals/soundtouch)
 
+enable_testing()
```
Comment (MerryMage) [inline CMakeLists.txt]: This line needs to be in the root `CMakeLists.txt` and be before `add_subdirectory(src)`.

**Comment by bunnei** [comment]
LGTM. hopefully people will write tests for this :p

**Comment by MerryMage** [comment]
\tests/
